packages/firestore/src/util/obj_map.ts

```diff
@@ -107,4 +107,64 @@ export class ObjectMap<KeyType extends Equatable<KeyType>, ValueType> {
   isEmpty(): boolean {
     return objUtil.isEmpty(this.inner);
   }
+
+  [Symbol.iterator](): Iterator<{ key: KeyType; value: ValueType }> {
+    // We don't care about the keys, all of the actual keys are in the
+    // arrays that are the values of the inner object.
+    const it = objUtil.values(this.inner)[Symbol.iterator]();
```

Comment:
This creates a copy of the value array, which is a little surprising for an iterator. I don't know a good way around this without using generators (which I assume we can't), so I am going to ignore this.

I do think this could be simplified quite a bit though. Wouldn't this work just as well:

```
    const values : Array<{ key: KeyType; value: ValueType }> = [];
    this.forEach((key, value) => values.push({key, value}));
    return values[Symbol.iterator]();
```

You might not need the ObjectMapIterator.

Reply:
Unfortunately, I don't know a way to avoid creating the array of values. Objects don't support iterators by default, and there's no ordering either to be able to start and stop iteration through the object.

As for using `forEach`, it kinda depends on whether or not we think there are collisions. `ObjectMapIterator` doesn't need to create all of the entries that will be returned at once. That's only a benefit though if there are collisions, since we do need the values array.

I'm not sure I have a strong opinion either way, since it depends on what's stored. Creating the whole array up front does have the advantage of not needing to export an extra `Entry` type, I suppose. 

Do you feel strongly either way?

Reply:
The two solutions (copy every value and copy every set of matching values) have such different complexities that I would much prefer to pick the simpler one.  
We can look at the usage in the SDK and see how common collisions are, but from memory it seems that they should be pretty rare (my IDE doesn't open at the moment...). The overhead from an occasional canonical ID collision can probably be ignored.

Reply:
Ok, I went with the `forEach` method.